MadeByMunstersLLC/fuel
# Feature category endpoints 1 3 17

**Changes:**
* Adds slugs to both tags and categories to follow along with the use of friendly_id gem as posts is using.
* Adds counter cache to categories on posts as this maybe needed and is a nice to have
* Adds `/categories.json` endpoint to get all categories back
* Adds `/posts.json?category=:category_slug` in order to get back all posts matching the category slug param. This includes using the existing pagination in place on the Posts#Index.

**Testing:**
* cd into `/spec/dummy`
* run `rake db:drop && rake db:create && rake db:migrate`
* add categories into the database
* check that `http://127.0.0.1:3000/blog/categories.json` brings you a json list of the categories
* add posts to the database with different categories attached.
* check that you can go to `http://127.0.0.1:3000/blog/posts.json?category=:your_category` and that you retrieve back the list of posts with that category.
* check that on the headers you are still seeing `Content-Range:0-1/2` for existing pagination
* check that inputing a empty or non-existent category returns back the default from `kaminari` gem which is a `204 No Content`.

**Notes:**
currently we only want to be able to search for posts by category although I went ahead and added slug support for tags incase we need to modify later we will already have slugs on all the tags.

The slug support was to make it easier on Kurt to be able to get a pre-defined category without first having to hit the categories endpoint to find the id.

**Trello Link:**
https://trello.com/c/z1rQdZqz/968-update-fuel-gem-to-allow-for-pagination-of-categories
